File: Problems 3100-3199/3108. Minimum Cost Walk in Weighted Graph/solution.py
class UnionFind:
    def __init__(self, n):
        self.parent = [i for i in range(n)]
        self.rank = [0]*n
        # two's complement of -1 is 1111111111
        self.cost = [-1]*n

    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    
    def union(self, x, y, weight):
        rootX = self.find(x)
        rootY = self.find(y)

        # Could have more than one edge between two same nodes
        # if rootX == rootY: return

        if self.rank[rootX] < self.rank[rootY]:
            rootX, rootY = rootY, rootX
        
        self.parent[rootY] = rootX
        if self.rank[rootX] == self.rank[rootY]:
            self.rank[rootX] += 1
        
        self.cost[rootX] = self.cost[rootX] & self.cost[rootY] & weight

    def minCost(self, x, y):
        if x == y: return 0
        rootX = self.find(x)
        rootY = self.find(y)

        if rootX != rootY:
            return -1
        
        return self.cost[rootX]

class Solution:
    def minimumCost(self, n: int, edges: List[List[int]], query: List[List[int]]) -> List[int]:
        # bitwising numbers will always result in <= result,
        # so we can just take bitwise and of the weights in
        # each group as the minimum
        # and each group means it is possible to visit each
        # node in that group since it's connected
        uf = UnionFind(n)
        for u, v, w in edges:
            uf.union(u, v, w)
        
        res = []
        for u, v in query:
            res.append(uf.minCost(u, v))
        
        return res

File: Problems 3100-3199/3108. Minimum Cost Walk in Weighted Graph/solution.cpp
class UnionFind {
private:
    // int n;
    vector<int> parent;
    vector<int> rank;
    vector<int> cost;
public:
    // -1 : 11111111 -> two's complement
    // UnionFind(int n) : n(n), parent(n), rank(n, 0), cost(n, -1) {
        UnionFind(int n) : parent(n), rank(n, 0), cost(n, -1) {
        for (int i = 0; i < n; i++) {
            parent[i] = i;
        }
    }

    int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }

    void unionRank(int x, int y, int weight) {
        int rootX = find(x);
        int rootY = find(y);

        if (rank[rootX] < rank[rootY]) swap(rootX, rootY);

        parent[rootY] = rootX;
        if (rank[rootX] == rank[rootY]) rank[rootX]++;

        cost[rootX] = cost[rootY] = cost[rootX] & cost[rootY] & weight;
    }

    int getMinCost(int x, int y) {
        if (x == y) return 0;
        int rootX = find(x);
        int rootY = find(y);

        if (rootX != rootY) return -1;

        return cost[rootX];
    }
};

class Solution {
public:
    vector<int> minimumCost(int n, vector<vector<int>>& edges, vector<vector<int>>& query) {
        // since we can walk same vertices any amount of time
        // we should find out connected components (which means)
        // we can traverse anywhere in that group of nodes
        // just take the bitwise and of the entire group
        // because it will always give us the minimum
        UnionFind uf(n);
        for (auto& edge : edges) {
            uf.unionRank(edge[0], edge[1], edge[2]);
        }

        vector<int> res;
        for (auto& q : query) {
            res.push_back(uf.getMinCost(q[0], q[1]));
        }

        return res;

    }
};